csharp/src/Ice/ObjectAdapter.cs

```diff
@@ -928,15 +823,16 @@ internal void DecDirectCount()
                 Debug.Assert(_directCount > 0);
                 if (--_directCount == 0)
                 {
-                    System.Threading.Monitor.PulseAll(_mutex);
+                    // Don't call SetResult directly to avoid continuations running synchronously
+                    Task.Run(() => _directDispatchCompletionSource?.SetResult(null));
                 }
             }
         }
 
         private void CheckForDeactivation()
         {
             // Must be called with _mutex locked.
```

Comment:
My preference is to simply relock the mutex.

Reply:
I removed this helper method and just throw ObjectDisposedException directly